// reka-ui/dist/index.js
import '@floating-ui/vue';
export { _ as VisuallyHidden } from './VisuallyHidden/VisuallyHidden.js';
export { _ as AccordionContent } from './Accordion/AccordionContent.js';
export { _ as AccordionHeader } from './Accordion/AccordionHeader.js';
export { _ as AccordionItem, i as injectAccordionItemContext } from './Accordion/AccordionItem.js';
export { _ as AccordionRoot, i as injectAccordionRootContext } from './Accordion/AccordionRoot.js';
export { _ as AccordionTrigger } from './Accordion/AccordionTrigger.js';
export { _ as AlertDialogAction } from './AlertDialog/AlertDialogAction.js';
export { _ as AlertDialogCancel } from './AlertDialog/AlertDialogCancel.js';
export { _ as AlertDialogContent, i as injectAlertDialogContentContext } from './AlertDialog/AlertDialogContent.js';
export { _ as AlertDialogDescription } from './AlertDialog/AlertDialogDescription.js';
export { _ as AlertDialogOverlay } from './AlertDialog/AlertDialogOverlay.js';
export { _ as AlertDialogPortal } from './AlertDialog/AlertDialogPortal.js';
export { _ as AlertDialogRoot } from './AlertDialog/AlertDialogRoot.js';
export { _ as AlertDialogTitle } from './AlertDialog/AlertDialogTitle.js';
export { _ as AlertDialogTrigger } from './AlertDialog/AlertDialogTrigger.js';
export { _ as AspectRatio } from './AspectRatio/AspectRatio.js';
export { _ as AvatarFallback } from './Avatar/AvatarFallback.js';
export { _ as AvatarImage } from './Avatar/AvatarImage.js';
export { _ as AvatarRoot, i as injectAvatarRootContext } from './Avatar/AvatarRoot.js';
export { _ as CalendarCell } from './Calendar/CalendarCell.js';
export { _ as CalendarCellTrigger } from './Calendar/CalendarCellTrigger.js';
export { _ as CalendarGrid } from './Calendar/CalendarGrid.js';
export { _ as CalendarGridBody } from './Calendar/CalendarGridBody.js';
export { _ as CalendarGridHead } from './Calendar/CalendarGridHead.js';
export { _ as CalendarGridRow } from './Calendar/CalendarGridRow.js';
export { _ as CalendarHeadCell } from './Calendar/CalendarHeadCell.js';
export { _ as CalendarHeader } from './Calendar/CalendarHeader.js';
export { _ as CalendarHeading } from './Calendar/CalendarHeading.js';
export { _ as CalendarNext } from './Calendar/CalendarNext.js';
export { _ as CalendarPrev } from './Calendar/CalendarPrev.js';
export { _ as CalendarRoot, i as injectCalendarRootContext } from './Calendar/CalendarRoot.js';
export { _ as CheckboxGroupRoot, i as injectCheckboxGroupRootContext } from './Checkbox/CheckboxGroupRoot.js';
export { _ as CheckboxIndicator } from './Checkbox/CheckboxIndicator.js';
export { _ as CheckboxRoot, i as injectCheckboxRootContext } from './Checkbox/CheckboxRoot.js';
export { _ as CollapsibleContent } from './Collapsible/CollapsibleContent.js';
export { _ as CollapsibleRoot, i as injectCollapsibleRootContext } from './Collapsible/CollapsibleRoot.js';
export { _ as CollapsibleTrigger } from './Collapsible/CollapsibleTrigger.js';
export { _ as ComboboxAnchor } from './Combobox/ComboboxAnchor.js';
export { _ as ComboboxArrow } from './Combobox/ComboboxArrow.js';
export { _ as ComboboxCancel } from './Combobox/ComboboxCancel.js';
export { _ as ComboboxContent } from './Combobox/ComboboxContent.js';
export { _ as ComboboxEmpty } from './Combobox/ComboboxEmpty.js';
export { _ as ComboboxGroup, i as injectComboboxGroupContext } from './Combobox/ComboboxGroup.js';
export { _ as ComboboxInput } from './Combobox/ComboboxInput.js';
export { _ as ComboboxItem } from './Combobox/ComboboxItem.js';
export { _ as ComboboxItemIndicator } from './Combobox/ComboboxItemIndicator.js';
export { _ as ComboboxLabel } from './Combobox/ComboboxLabel.js';
export { _ as ComboboxPortal } from './Combobox/ComboboxPortal.js';
export { _ as ComboboxRoot, i as injectComboboxRootContext } from './Combobox/ComboboxRoot.js';
export { _ as ComboboxSeparator } from './Combobox/ComboboxSeparator.js';
export { _ as ComboboxTrigger } from './Combobox/ComboboxTrigger.js';
export { _ as ComboboxViewport } from './Combobox/ComboboxViewport.js';
export { _ as ComboboxVirtualizer } from './Combobox/ComboboxVirtualizer.js';
export { _ as ConfigProvider, i as injectConfigProviderContext } from './ConfigProvider/ConfigProvider.js';
export { _ as ContextMenuArrow } from './ContextMenu/ContextMenuArrow.js';
export { _ as ContextMenuCheckboxItem } from './ContextMenu/ContextMenuCheckboxItem.js';
export { _ as ContextMenuContent } from './ContextMenu/ContextMenuContent.js';
export { _ as ContextMenuGroup } from './ContextMenu/ContextMenuGroup.js';
export { _ as ContextMenuItem } from './ContextMenu/ContextMenuItem.js';
export { _ as ContextMenuItemIndicator } from './ContextMenu/ContextMenuItemIndicator.js';
export { _ as ContextMenuLabel } from './ContextMenu/ContextMenuLabel.js';
export { _ as ContextMenuPortal } from './ContextMenu/ContextMenuPortal.js';
export { _ as ContextMenuRadioGroup } from './ContextMenu/ContextMenuRadioGroup.js';
export { _ as ContextMenuRadioItem } from './ContextMenu/ContextMenuRadioItem.js';
export { _ as ContextMenuRoot, i as injectContextMenuRootContext } from './ContextMenu/ContextMenuRoot.js';
export { _ as ContextMenuSeparator } from './ContextMenu/ContextMenuSeparator.js';
export { _ as ContextMenuSub } from './ContextMenu/ContextMenuSub.js';
export { _ as ContextMenuSubContent } from './ContextMenu/ContextMenuSubContent.js';
export { _ as ContextMenuSubTrigger } from './ContextMenu/ContextMenuSubTrigger.js';
export { _ as ContextMenuTrigger } from './ContextMenu/ContextMenuTrigger.js';
export { _ as DateFieldInput } from './DateField/DateFieldInput.js';
export { _ as DateFieldRoot, i as injectDateFieldRootContext } from './DateField/DateFieldRoot.js';
import { defineComponent, createBlock, openBlock, unref, normalizeProps, guardReactiveProps, withCtx, renderSlot } from 'vue';
import { _ as _sfc_main$1 } from './Popover/PopoverAnchor.js';
export { _ as DatePickerArrow } from './DatePicker/DatePickerArrow.js';
export { _ as DatePickerCalendar } from './DatePicker/DatePickerCalendar.js';
export { _ as DatePickerCell } from './DatePicker/DatePickerCell.js';
export { _ as DatePickerCellTrigger } from './DatePicker/DatePickerCellTrigger.js';
export { _ as DatePickerClose } from './DatePicker/DatePickerClose.js';
export { _ as DatePickerContent } from './DatePicker/DatePickerContent.js';
export { _ as DatePickerField } from './DatePicker/DatePickerField.js';
export { _ as DatePickerGrid } from './DatePicker/DatePickerGrid.js';
export { _ as DatePickerGridBody } from './DatePicker/DatePickerGridBody.js';
export { _ as DatePickerGridHead } from './DatePicker/DatePickerGridHead.js';
export { _ as DatePickerGridRow } from './DatePicker/DatePickerGridRow.js';
export { _ as DatePickerHeadCell } from './DatePicker/DatePickerHeadCell.js';
export { _ as DatePickerHeader } from './DatePicker/DatePickerHeader.js';
export { _ as DatePickerHeading } from './DatePicker/DatePickerHeading.js';
export { _ as DatePickerInput } from './DatePicker/DatePickerInput.js';
export { _ as DatePickerNext } from './DatePicker/DatePickerNext.js';
export { _ as DatePickerPrev } from './DatePicker/DatePickerPrev.js';
export { _ as DatePickerRoot, i as injectDatePickerRootContext } from './DatePicker/DatePickerRoot.js';
export { _ as DatePickerTrigger } from './DatePicker/DatePickerTrigger.js';
export { _ as DateRangeFieldInput } from './DateRangeField/DateRangeFieldInput.js';
export { _ as DateRangeFieldRoot, i as injectDateRangeFieldRootContext } from './DateRangeField/DateRangeFieldRoot.js';
export { _ as DateRangePickerAnchor } from './DateRangePicker/DateRangePickerAnchor.js';
export { _ as DateRangePickerArrow } from './DateRangePicker/DateRangePickerArrow.js';
export { _ as DateRangePickerCalendar } from './DateRangePicker/DateRangePickerCalendar.js';
export { _ as DateRangePickerCell } from './DateRangePicker/DateRangePickerCell.js';
export { _ as DateRangePickerCellTrigger } from './DateRangePicker/DateRangePickerCellTrigger.js';
export { _ as DateRangePickerClose } from './DateRangePicker/DateRangePickerClose.js';
export { _ as DateRangePickerContent } from './DateRangePicker/DateRangePickerContent.js';
export { _ as DateRangePickerField } from './DateRangePicker/DateRangePickerField.js';
export { _ as DateRangePickerGrid } from './DateRangePicker/DateRangePickerGrid.js';
export { _ as DateRangePickerGridBody } from './DateRangePicker/DateRangePickerGridBody.js';
export { _ as DateRangePickerGridHead } from './DateRangePicker/DateRangePickerGridHead.js';
export { _ as DateRangePickerGridRow } from './DateRangePicker/DateRangePickerGridRow.js';
export { _ as DateRangePickerHeadCell } from './DateRangePicker/DateRangePickerHeadCell.js';
export { _ as DateRangePickerHeader } from './DateRangePicker/DateRangePickerHeader.js';
export { _ as DateRangePickerHeading } from './DateRangePicker/DateRangePickerHeading.js';
export { _ as DateRangePickerInput } from './DateRangePicker/DateRangePickerInput.js';
export { _ as DateRangePickerNext } from './DateRangePicker/DateRangePickerNext.js';
export { _ as DateRangePickerPrev } from './DateRangePicker/DateRangePickerPrev.js';
export { _ as DateRangePickerRoot, i as injectDateRangePickerRootContext } from './DateRangePicker/DateRangePickerRoot.js';
export { _ as DateRangePickerTrigger } from './DateRangePicker/DateRangePickerTrigger.js';
export { _ as DialogClose } from './Dialog/DialogClose.js';
export { _ as DialogContent } from './Dialog/DialogContent.js';
export { _ as DialogDescription } from './Dialog/DialogDescription.js';
export { _ as DialogOverlay } from './Dialog/DialogOverlay.js';
export { _ as DialogPortal } from './Dialog/DialogPortal.js';
export { _ as DialogRoot, i as injectDialogRootContext } from './Dialog/DialogRoot.js';
export { _ as DialogTitle } from './Dialog/DialogTitle.js';
export { _ as DialogTrigger } from './Dialog/DialogTrigger.js';
export { _ as DropdownMenuArrow } from './DropdownMenu/DropdownMenuArrow.js';
export { _ as DropdownMenuCheckboxItem } from './DropdownMenu/DropdownMenuCheckboxItem.js';
export { _ as DropdownMenuContent } from './DropdownMenu/DropdownMenuContent.js';
export { _ as DropdownMenuGroup } from './DropdownMenu/DropdownMenuGroup.js';
export { _ as DropdownMenuItem } from './DropdownMenu/DropdownMenuItem.js';
export { _ as DropdownMenuItemIndicator } from './DropdownMenu/DropdownMenuItemIndicator.js';
export { _ as DropdownMenuLabel } from './DropdownMenu/DropdownMenuLabel.js';
export { _ as DropdownMenuPortal } from './DropdownMenu/DropdownMenuPortal.js';
export { _ as DropdownMenuRadioGroup } from './DropdownMenu/DropdownMenuRadioGroup.js';
export { _ as DropdownMenuRadioItem } from './DropdownMenu/DropdownMenuRadioItem.js';
export { _ as DropdownMenuRoot, i as injectDropdownMenuRootContext } from './DropdownMenu/DropdownMenuRoot.js';
export { _ as DropdownMenuSeparator } from './DropdownMenu/DropdownMenuSeparator.js';
export { _ as DropdownMenuSub } from './DropdownMenu/DropdownMenuSub.js';
export { _ as DropdownMenuSubContent } from './DropdownMenu/DropdownMenuSubContent.js';
export { _ as DropdownMenuSubTrigger } from './DropdownMenu/DropdownMenuSubTrigger.js';
export { _ as DropdownMenuTrigger } from './DropdownMenu/DropdownMenuTrigger.js';
export { _ as EditableArea } from './Editable/EditableArea.js';
export { _ as EditableCancelTrigger } from './Editable/EditableCancelTrigger.js';
export { _ as EditableEditTrigger } from './Editable/EditableEditTrigger.js';
export { _ as EditableInput } from './Editable/EditableInput.js';
export { _ as EditablePreview } from './Editable/EditablePreview.js';
export { _ as EditableRoot, i as injectEditableRootContext } from './Editable/EditableRoot.js';
export { _ as EditableSubmitTrigger } from './Editable/EditableSubmitTrigger.js';
export { _ as FocusScope } from './FocusScope/FocusScope.js';
export { _ as HoverCardArrow } from './HoverCard/HoverCardArrow.js';
export { _ as HoverCardContent } from './HoverCard/HoverCardContent.js';
export { _ as HoverCardPortal } from './HoverCard/HoverCardPortal.js';
export { _ as HoverCardRoot, i as injectHoverCardRootContext } from './HoverCard/HoverCardRoot.js';
export { _ as HoverCardTrigger } from './HoverCard/HoverCardTrigger.js';
export { _ as Label } from './Label/Label.js';
export { _ as ListboxContent } from './Listbox/ListboxContent.js';
export { _ as ListboxFilter } from './Listbox/ListboxFilter.js';
export { _ as ListboxGroup, i as injectListboxGroupContext } from './Listbox/ListboxGroup.js';
export { _ as ListboxGroupLabel } from './Listbox/ListboxGroupLabel.js';
export { _ as ListboxItem, i as injectComboboxItemContext, i as injectListboxItemContext } from './Listbox/ListboxItem.js';
export { _ as ListboxItemIndicator } from './Listbox/ListboxItemIndicator.js';
export { _ as ListboxRoot, i as injectListboxRootContext } from './Listbox/ListboxRoot.js';
export { _ as ListboxVirtualizer } from './Listbox/ListboxVirtualizer.js';
export { _ as MenubarArrow } from './Menubar/MenubarArrow.js';
export { _ as MenubarCheckboxItem } from './Menubar/MenubarCheckboxItem.js';
export { _ as MenubarContent } from './Menubar/MenubarContent.js';
export { _ as MenubarGroup } from './Menubar/MenubarGroup.js';
export { _ as MenubarItem } from './Menubar/MenubarItem.js';
export { _ as MenubarItemIndicator } from './Menubar/MenubarItemIndicator.js';
export { _ as MenubarLabel } from './Menubar/MenubarLabel.js';
export { _ as MenubarMenu, i as injectMenubarMenuContext } from './Menubar/MenubarMenu.js';
export { _ as MenubarPortal } from './Menubar/MenubarPortal.js';
export { _ as MenubarRadioGroup } from './Menubar/MenubarRadioGroup.js';
export { _ as MenubarRadioItem } from './Menubar/MenubarRadioItem.js';
export { _ as MenubarRoot, i as injectMenubarRootContext } from './Menubar/MenubarRoot.js';
export { _ as MenubarSeparator } from './Menubar/MenubarSeparator.js';
export { _ as MenubarSub } from './Menubar/MenubarSub.js';
export { _ as MenubarSubContent } from './Menubar/MenubarSubContent.js';
export { _ as MenubarSubTrigger } from './Menubar/MenubarSubTrigger.js';
export { _ as MenubarTrigger } from './Menubar/MenubarTrigger.js';
export { _ as NavigationMenuContent } from './NavigationMenu/NavigationMenuContent.js';
export { _ as NavigationMenuIndicator } from './NavigationMenu/NavigationMenuIndicator.js';
export { _ as NavigationMenuItem, i as injectNavigationMenuItemContext } from './NavigationMenu/NavigationMenuItem.js';
export { _ as NavigationMenuLink } from './NavigationMenu/NavigationMenuLink.js';
export { _ as NavigationMenuList } from './NavigationMenu/NavigationMenuList.js';
export { _ as NavigationMenuRoot, i as injectNavigationMenuContext } from './NavigationMenu/NavigationMenuRoot.js';
export { _ as NavigationMenuSub } from './NavigationMenu/NavigationMenuSub.js';
export { _ as NavigationMenuTrigger } from './NavigationMenu/NavigationMenuTrigger.js';
export { _ as NavigationMenuViewport } from './NavigationMenu/NavigationMenuViewport.js';
export { _ as NumberFieldDecrement } from './NumberField/NumberFieldDecrement.js';
export { _ as NumberFieldIncrement } from './NumberField/NumberFieldIncrement.js';
export { _ as NumberFieldInput } from './NumberField/NumberFieldInput.js';
export { _ as NumberFieldRoot, i as injectNumberFieldRootContext } from './NumberField/NumberFieldRoot.js';
export { _ as PaginationEllipsis } from './Pagination/PaginationEllipsis.js';
export { _ as PaginationFirst } from './Pagination/PaginationFirst.js';
export { _ as PaginationLast } from './Pagination/PaginationLast.js';
export { _ as PaginationList } from './Pagination/PaginationList.js';
export { _ as PaginationListItem } from './Pagination/PaginationListItem.js';
export { _ as PaginationNext } from './Pagination/PaginationNext.js';
export { _ as PaginationPrev } from './Pagination/PaginationPrev.js';
export { _ as PaginationRoot, i as injectPaginationRootContext } from './Pagination/PaginationRoot.js';
export { _ as PinInputInput } from './PinInput/PinInputInput.js';
export { _ as PinInputRoot, i as injectPinInputRootContext } from './PinInput/PinInputRoot.js';
export { _ as PopoverArrow } from './Popover/PopoverArrow.js';
export { _ as PopoverClose } from './Popover/PopoverClose.js';
export { _ as PopoverContent } from './Popover/PopoverContent.js';
export { _ as PopoverPortal } from './Popover/PopoverPortal.js';
export { _ as PopoverRoot, i as injectPopoverRootContext } from './Popover/PopoverRoot.js';
export { _ as PopoverTrigger } from './Popover/PopoverTrigger.js';
export { _ as ProgressIndicator } from './Progress/ProgressIndicator.js';
export { _ as ProgressRoot, i as injectProgressRootContext } from './Progress/ProgressRoot.js';
export { _ as RadioGroupIndicator } from './RadioGroup/RadioGroupIndicator.js';
export { _ as RadioGroupItem, i as injectRadioGroupItemContext } from './RadioGroup/RadioGroupItem.js';
export { _ as RadioGroupRoot, i as injectRadioGroupRootContext } from './RadioGroup/RadioGroupRoot.js';
export { _ as RangeCalendarCell } from './RangeCalendar/RangeCalendarCell.js';
export { _ as RangeCalendarCellTrigger } from './RangeCalendar/RangeCalendarCellTrigger.js';
export { _ as RangeCalendarGrid } from './RangeCalendar/RangeCalendarGrid.js';
export { _ as RangeCalendarGridBody } from './RangeCalendar/RangeCalendarGridBody.js';
export { _ as RangeCalendarGridHead } from './RangeCalendar/RangeCalendarGridHead.js';
export { _ as RangeCalendarGridRow } from './RangeCalendar/RangeCalendarGridRow.js';
export { _ as RangeCalendarHeadCell } from './RangeCalendar/RangeCalendarHeadCell.js';
export { _ as RangeCalendarHeader } from './RangeCalendar/RangeCalendarHeader.js';
export { _ as RangeCalendarHeading } from './RangeCalendar/RangeCalendarHeading.js';
export { _ as RangeCalendarNext } from './RangeCalendar/RangeCalendarNext.js';
export { _ as RangeCalendarPrev } from './RangeCalendar/RangeCalendarPrev.js';
export { _ as RangeCalendarRoot, i as injectRangeCalendarRootContext } from './RangeCalendar/RangeCalendarRoot.js';
export { _ as RovingFocusGroup } from './RovingFocus/RovingFocusGroup.js';
export { _ as RovingFocusItem } from './RovingFocus/RovingFocusItem.js';
export { _ as ScrollAreaCorner } from './ScrollArea/ScrollAreaCorner.js';
export { _ as ScrollAreaRoot, i as injectScrollAreaRootContext } from './ScrollArea/ScrollAreaRoot.js';
export { _ as ScrollAreaScrollbar, i as injectScrollAreaScrollbarContext } from './ScrollArea/ScrollAreaScrollbar.js';
export { _ as ScrollAreaThumb } from './ScrollArea/ScrollAreaThumb.js';
export { _ as ScrollAreaViewport } from './ScrollArea/ScrollAreaViewport.js';
export { _ as SelectArrow } from './Select/SelectArrow.js';
export { _ as SelectContent } from './Select/SelectContent.js';
export { _ as SelectGroup, i as injectSelectGroupContext } from './Select/SelectGroup.js';
export { _ as SelectIcon } from './Select/SelectIcon.js';
export { _ as SelectItem, i as injectSelectItemContext } from './Select/SelectItem.js';
export { _ as SelectItemIndicator } from './Select/SelectItemIndicator.js';
export { _ as SelectItemText } from './Select/SelectItemText.js';
export { _ as SelectLabel } from './Select/SelectLabel.js';
export { _ as SelectPortal } from './Select/SelectPortal.js';
export { _ as SelectRoot, i as injectSelectRootContext } from './Select/SelectRoot.js';
export { _ as SelectScrollDownButton } from './Select/SelectScrollDownButton.js';
export { _ as SelectScrollUpButton } from './Select/SelectScrollUpButton.js';
export { _ as SelectSeparator } from './Select/SelectSeparator.js';
export { _ as SelectTrigger } from './Select/SelectTrigger.js';
export { _ as SelectValue } from './Select/SelectValue.js';
export { _ as SelectViewport } from './Select/SelectViewport.js';
export { _ as Separator } from './Separator/Separator.js';
export { _ as SliderRange } from './Slider/SliderRange.js';
export { _ as SliderRoot, i as injectSliderRootContext } from './Slider/SliderRoot.js';
export { _ as SliderThumb } from './Slider/SliderThumb.js';
export { _ as SliderTrack } from './Slider/SliderTrack.js';
export { _ as SplitterGroup, i as injectSplitterGroupContext } from './Splitter/SplitterGroup.js';
export { _ as SplitterPanel } from './Splitter/SplitterPanel.js';
export { _ as SplitterResizeHandle } from './Splitter/SplitterResizeHandle.js';
export { _ as StepperDescription } from './Stepper/StepperDescription.js';
export { _ as StepperIndicator } from './Stepper/StepperIndicator.js';
export { _ as StepperItem, i as injectStepperItemContext } from './Stepper/StepperItem.js';
export { _ as StepperRoot, i as injectStepperRootContext } from './Stepper/StepperRoot.js';
export { _ as StepperSeparator } from './Stepper/StepperSeparator.js';
export { _ as StepperTitle } from './Stepper/StepperTitle.js';
export { _ as StepperTrigger } from './Stepper/StepperTrigger.js';
export { _ as SwitchRoot, i as injectSwitchRootContext } from './Switch/SwitchRoot.js';
export { _ as SwitchThumb } from './Switch/SwitchThumb.js';
export { _ as TabsContent } from './Tabs/TabsContent.js';
export { _ as TabsIndicator } from './Tabs/TabsIndicator.js';
export { _ as TabsList } from './Tabs/TabsList.js';
export { _ as TabsRoot, i as injectTabsRootContext } from './Tabs/TabsRoot.js';
export { _ as TabsTrigger } from './Tabs/TabsTrigger.js';
export { _ as TagsInputClear } from './TagsInput/TagsInputClear.js';
export { _ as TagsInputInput } from './TagsInput/TagsInputInput.js';
export { _ as TagsInputItem, i as injectTagsInputItemContext } from './TagsInput/TagsInputItem.js';
export { _ as TagsInputItemDelete } from './TagsInput/TagsInputItemDelete.js';
export { _ as TagsInputItemText } from './TagsInput/TagsInputItemText.js';
export { _ as TagsInputRoot, i as injectTagsInputRootContext } from './TagsInput/TagsInputRoot.js';
export { _ as TimeFieldInput } from './TimeField/TimeFieldInput.js';
export { _ as TimeFieldRoot, i as injectTimeFieldRootContext } from './TimeField/TimeFieldRoot.js';
export { _ as ToastAction } from './Toast/ToastAction.js';
export { _ as ToastClose } from './Toast/ToastClose.js';
export { _ as ToastDescription } from './Toast/ToastDescription.js';
export { _ as ToastPortal } from './Toast/ToastPortal.js';
export { _ as ToastProvider, i as injectToastProviderContext } from './Toast/ToastProvider.js';
export { _ as ToastRoot } from './Toast/ToastRoot.js';
export { _ as ToastTitle } from './Toast/ToastTitle.js';
export { _ as ToastViewport } from './Toast/ToastViewport.js';
export { _ as Toggle } from './Toggle/Toggle.js';
export { _ as ToggleGroupItem } from './ToggleGroup/ToggleGroupItem.js';
export { _ as ToggleGroupRoot, i as injectToggleGroupRootContext } from './ToggleGroup/ToggleGroupRoot.js';
export { _ as ToolbarButton } from './Toolbar/ToolbarButton.js';
export { _ as ToolbarLink } from './Toolbar/ToolbarLink.js';
export { _ as ToolbarRoot, i as injectToolbarRootContext } from './Toolbar/ToolbarRoot.js';
export { _ as ToolbarSeparator } from './Toolbar/ToolbarSeparator.js';
export { _ as ToolbarToggleGroup } from './Toolbar/ToolbarToggleGroup.js';
export { _ as ToolbarToggleItem } from './Toolbar/ToolbarToggleItem.js';
export { _ as TooltipArrow } from './Tooltip/TooltipArrow.js';
export { _ as TooltipContent } from './Tooltip/TooltipContent.js';
export { _ as TooltipPortal } from './Tooltip/TooltipPortal.js';
export { _ as TooltipProvider, i as injectTooltipProviderContext } from './Tooltip/TooltipProvider.js';
export { _ as TooltipRoot, i as injectTooltipRootContext } from './Tooltip/TooltipRoot.js';
export { _ as TooltipTrigger } from './Tooltip/TooltipTrigger.js';
export { _ as TreeItem } from './Tree/TreeItem.js';
export { _ as TreeRoot, i as injectTreeRootContext } from './Tree/TreeRoot.js';
export { _ as TreeVirtualizer } from './Tree/TreeVirtualizer.js';
export { _ as Viewport } from './Viewport/Viewport.js';
export { P as Primitive } from './Primitive/Primitive.js';
export { S as Slot } from './Primitive/Slot.js';
export { c as createContext } from './shared/createContext.js';
export { u as useBodyScrollLock } from './shared/useBodyScrollLock.js';
export { u as useDateFormatter } from './shared/useDateFormatter.js';
export { u as useEmitAsProps } from './shared/useEmitAsProps.js';
export { u as useFilter } from './shared/useFilter.js';
export { u as useForwardExpose } from './shared/useForwardExpose.js';
export { u as useForwardProps } from './shared/useForwardProps.js';
export { u as useForwardPropsEmits } from './shared/useForwardPropsEmits.js';
export { u as useId } from './shared/useId.js';
export { u as useStateMachine } from './shared/useStateMachine.js';
export { w as withDefault } from './shared/withDefault.js';
export { P as Presence } from './Presence/Presence.js';

const _sfc_main = /* @__PURE__ */ defineComponent({
  __name: "DatePickerAnchor",
  props: {
    reference: {},
    asChild: { type: Boolean },
    as: {}
  },
  setup(__props) {
    const props = __props;
    return (_ctx, _cache) => {
      return openBlock(), createBlock(unref(_sfc_main$1), normalizeProps(guardReactiveProps(props)), {
        default: withCtx(() => [
          renderSlot(_ctx.$slots, "default")
        ]),
        _: 3
      }, 16);
    };
  }
});

export { _sfc_main as DatePickerAnchor, _sfc_main$1 as PopoverAnchor };
